oslo/transformers/oslo_init.py

```diff
@@ -0,0 +1,451 @@
+
+        if not self.is_exist('expert_parallelism') or not self.expert_parallelism["enable"]:
+            self.expert_parallelism = None
+        else:
+            if self.expert_parallelism['parallel_size'] is None:
+                log_dist(
+                    "expert_parallelism can not be usable because parallel_size is required."
+                    , logging.WARNING
+                )
+                self.expert_parallelism = None
+            else:
+                log_dist(
+                    f"expert_parallelism: enabled\n\tparallel_size: {self.expert_parallelism['parallel_size']}"
+                )
+        # print(str(self.__dict__.items()))
+
+
+def init_oslo_features(
+    oslo_init_config: OsloTrainerConfig,) -> Tuple[ParallelContext, List]:
+    """
+    Init OSLO features with json or dict configuration user passed.
+    ParallelContext or other effective features should be defined on this function
+    and Trainer could use this outputs
+
+    This function returns two object, ParallelContext and WrapperModule from user config
+    TrainArgumet class use this to re-define model
+    >> model = ...
+    >> parallel_context = ParallelContext.from_torch(...)
+    >> wrapper_model = TensorParallel(model, parallel_context)
+    >> allocate_params(wrapper_model, parallel_context)
+    """
+    cfg = oslo_init_config
+    data_parallel_size = cfg.data_parallelism.parallel_size if cfg.data_parallelism else 1
+    sequence_parallel_size = cfg.sequence_parallelism.parallel_size if cfg.sequence_parallelism else 1
+    expert_parallel_size = cfg.expert_parallelism.parallel_size if cfg.expert_parallelism else 1
+    pipeline_parallel_size = cfg.pipeline_parallelism.parallel_size if cfg.pipeline_parallelism else 1
+    tensor_parallel_size, tensor_parallel_depth, tensor_parallel_mode = 1, 1, TENSOR_PARALLEL_MAPPING["1d"]
+    if cfg.tensor_parallelism:
+        tensor_parallel_size = cfg.tensor_parallelism.parallel_size
+        tensor_parallel_mode = TENSOR_PARALLEL_MAPPING[cfg.tensor_parallelism.parallel_mode]
+        if cfg.tensor_parallelism.is_exist('param'):
+            tensor_parallel_depth = cfg.tensor_parallelism.param['parallel_depth_2.5d']
+
+    parallel_context = ParallelContext.from_torch(
+        data_parallel_size=data_parallel_size,
+        sequence_parallel_size=sequence_parallel_size,
+        expert_parallel_size=expert_parallel_size,
+        pipeline_parallel_size=pipeline_parallel_size,
+        tensor_parallel_size=tensor_parallel_size,
+        tensor_parallel_depth=tensor_parallel_depth,
+        tensor_parallel_mode=tensor_parallel_mode
+    )
+
+    if tensor_parallel_size > 1 and sequence_parallel_size > 1:
+        raise AttributeError(
+            "TensorParallel and SequenceParallel can't be used at the same time. Modify oslo config to avoid wrong parallel setting"
+        )
+
+    model_wrapper = []
+
+    if data_parallel_size > 1:
+        if cfg.data_parallelism.zero_stage == 0:
```

Comment:
You don't need to do it in Trainer. It'll be handled in DataParallel module

Reply:
Ok. removed